### DISCUSSION TITLE
项目是很老的项目，最近升级为umi4，项目中不仅在用dva还混用了redux，升级后的redux里的state和dva里的state会冲突。。。

### DISCUSSION THREAD

**Discussion by kaixinobaoo** [umijs/umi#12309] [Q&A]
项目从umi2升到umi4。。layout是自定义的，想知道把props.children替换为<Outlet />后，想知道之前的redux里的state如何获取？目前这种写法导致渲染的组件拿到的state只有之前的redux里的state且报错dva中的state找不到。
`<ProLayout {...customProps}>
 <Provider store={store}>
        <div className={styles.contentWrap}>
              <Outlet />
         </div>
</Provider>
<ProLayout>
`

**Comment by fz6m** [comment]
1. 如果差距太大，迁移的时候最好创建一个新的 umi 4 项目，一点一点往里挪，一个一个解决问题，不然一次性报错太多不知如何下手解决。

2. umi 4 的 max 有一个内置 layout 菜单功能，如果不打开 `layout: {}` 就默认没有 layout ，所以把你的自定义 layout 放在 `src/layouts/index.tsx` 即可。

3. 关于 dva 的问题，我不了解 umi 2 。如果你用的是 umi 2 的 dva 插件，你得确认下以前 dva 的编写约定和现在 umi 4 提供的 dva 插件里的约定是不是一样的，如果有区别，就需要调整，文档在 [umi 4 dva](https://umijs.org/docs/max/dva) ；如果你的 dva 是自己接进去的，就没必要用 umi 4 的 dva 插件，以前怎么接入项目的继续保留即可。

4. 关于 redux 的问题，应该你只要确保不和 dva 冲突即可，旧项目应该是自己用 redux context 去使用的吧，先把 dva 的运行跑通了再处理 redux 的问题吧，还需要考虑下 redux 新旧版本是不是一样等问题。

如果旧项目 layout 是完全自己写的，就没必要迁就 umi 4 的 layout ，路由信息可以从 `useAppData` 里取，参考 `src/.umi/plugin-layout/*` 下面的文件 copy 了自己改下吧。

**Reply by kaixinobaoo** [reply]
layout问题我通过其他issue已经解决了。

 redux的问题是通过使用`<Outlet />`渲染的组件获取不到`<Provider store={store}> <div className={styles.contentWrap}> <Outlet /> </div> </Provider>`里的**store**，且如果添加了`<Provider store={store}> ...  </Provider>`这行包裹，通过layout渲染的所有子组件mapStateToProps中的state打印出来只有用store管理的数据而没有那些使用dva管理的状态数据。。。尝试过examples里的解决方案在入口导出rootContanier但并没有效果。
![image](https://github.com/umijs/umi/assets/167517494/ab3f21a8-a708-4934-aa78-5e1bf866c0fa)
而将layout中`<Provider store={store}> ...  </Provider>`这行包裹注释，mapStateToProps中的state打印出来则是用dva管理的状态数据。。不是很明白为什么会这样。。。我也想过干脆直接在layout渲染子组件的时候做区分，想问问你有什么更好的解决方案吗。。。

**Reply by fz6m** [reply]
你确定 dva 能和 redux 独立使用去混用吗，dva 本来就依赖于 redux ，用 `pnpm why react-redux` 和 `pnpm why redux` 看看你有几个实例吧，可能 dva 的 redux 和你独立使用的 redux 不是一个实例，如果有多实例，就可能会发生只能取到一边的数据。

如果有多实例，就得确保 dva 用的 redux 版本和你自己安装的是一个版本，并且在 alias 加一下，强确保 redux 的这俩依赖的唯一性，之后再试试吧。
